---
title: "Analysis of mass spectrometry data for 1st large library"
author: "Ute Hoffmann (Science For Life Laboratory (KTH), Stockholm, Sweden)"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    fig_width: 15
    fig_height: 8
    theme: united
    toc: false
    number_sections: true
  pdf_document:
    toc: yes
    number_sections: true
---



```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "~/Documents/bioinformatic_analyses/2024-05_LargeRubiscoLibrary/raw_data/MS_firstCultivation")
library(ggplot2)
library(tidyverse)
library(colorblindr)
library(ggrepel)
library(clusterProfiler)
library(enrichplot)
```

# Objective of analysis

Mass spectrometry data was collected to check if *Synechocystis* Rubisco (RbcL, RbcS) remains repressed after 3.7 and 8.6 generations and if dCas9 remains induced. The following samples were collected (compare also 2024-04-24_MSsamples.ods):
Quadruplicates of
  - 1% CO2, air, generation 0/4/8 (actual generations: 0, 3.8, 7.3)
  - 5% CO2, N2, generation 0/4/8 (actual generations: 0, 3.7, 8.6)

# Read in data

```{r, message=FALSE, warning=FALSE}
protein_slr_ID <- read.delim("2024-01-25_Protein_to_slrIdentifier.tsv", sep="\t", header=T)
names(protein_slr_ID) <- c("locus_tag", "Entry")

highCO2_data <- read.delim("dia_msstats_20240614_highco2_pcc6803.tsv", header=TRUE, dec=".", sep="\t")
highCO2_data <- left_join(highCO2_data, protein_slr_ID)
highCO2_data <- subset(highCO2_data, highCO2_data$ambig == "no" & !is.na(highCO2_data$log2FC) & !is.na(highCO2_data$adj.pvalue) & is.finite(highCO2_data$log2FC) & is.finite(highCO2_data$adj.pvalue))
lowCO2_comparison <- read.delim("dia_msstats_20240614_lowco2_pcc6803.tsv", header=TRUE, dec=".", sep="\t")
lowCO2_comparison <- left_join(lowCO2_comparison, protein_slr_ID)
lowCO2_comparison <- subset(lowCO2_comparison, lowCO2_comparison$ambig == "no" & !is.na(lowCO2_comparison$log2FC) & !is.na(lowCO2_comparison$adj.pvalue) & is.finite(lowCO2_comparison$log2FC) & is.finite(lowCO2_comparison$adj.pvalue))
```

Separate Tables by comparison:

```{r}
columns <- c("log2FC", "SE", "Tvalue", "DF", "pvalue", "adj.pvalue", "issue", "MissingPercentage", "Entry", "ambig", "gene_names_ordered_locus", "gene_names_primary", "Protein.names", "locus_tag")
highCO2_0_4 <- subset(highCO2_data, highCO2_data$Label=="highco2_4")
highCO2_0_8 <- subset(highCO2_data, highCO2_data$Label=="highco2_8")
write_csv(highCO2_0_4[,columns], "tables/MS_data_comparison_highCO2_gen0-4.csv")
write_csv(highCO2_0_8[,columns], "tables/MS_data_comparison_highCO2_gen0-8")

lowCO2_0_4 <- subset(lowCO2_comparison, lowCO2_comparison$Label=="lowco2_4")
lowCO2_0_8 <- subset(lowCO2_comparison, lowCO2_comparison$Label=="lowco2_8")
write_csv(subset(lowCO2_comparison, lowCO2_comparison$Label=="lowco2_4")[,columns], "tables/MS_data_comparison_lowCO2_gen0-4.csv")
write_csv(subset(lowCO2_comparison, lowCO2_comparison$Label=="lowco2_8")[,columns], "tables/MS_data_comparison_lowCO2_gen0-8.csv.csv")
```

# Plot Volcano plots

```{r, echo=FALSE}
volcanoPlot_ggplot <-  function(dframe, foldchange=1, padjusted=0.01, color=TRUE, text=FALSE, numbers=TRUE, lines_padj_FC=TRUE, colours=c("#e69f00b2","#005a96b2",  "#d3d3d3b2"))
{
  # compare: https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html
  
  # The significantly differentially expressed genes are the ones found in the upper-left and upper-right corners.
  # Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2FoldChange respectively positive or negative)
  
  # add a column of NAs
  df_copy <- as.data.frame(dframe)
  df_copy$diffexpressed <- "NO"
  df_copy$diffexpressed[df_copy$log2FC > foldchange & df_copy$adj.pvalue < padjusted] <- "UP"
  df_copy$diffexpressed[df_copy$log2FC < -foldchange & df_copy$adj.pvalue < padjusted] <- "DOWN"
  
  df_copy$diffexpressed[df_copy$Entry=="P54205"] <- "rbcLXS"
  df_copy$diffexpressed[df_copy$Entry=="P54206"] <- "rbcLXS"
  df_copy$diffexpressed[df_copy$Entry=="WTGAL"] <- "rbcLXS"
  df_copy$diffexpressed[df_copy$Entry=="DCAS9"] <- "rbcLXS"
  df_copy$gene_names_ordered_locus[df_copy$Entry=="DCAS9"] <- "dCas9"
  
  # prepare labels of plot
  df_copy$delabel <- NA
  df_copy$delabel[df_copy$diffexpressed != "NO"] <- df_copy$gene_names_ordered_locus[df_copy$diffexpressed != "NO"]
  df_copy$delabel[df_copy$Entry=="WTGAL"] <- "cbbM"
  # Plot
  p <- ggplot(data=df_copy, aes(x=log2FC, y=-log10(adj.pvalue), col=diffexpressed, label=delabel)) + geom_point(alpha=0.3, show.legend = FALSE) + 
    theme_light() + labs(y="-Log10(p.adj)", x="Log2FC") + theme(legend.position = "none")
  
  # color points
  mycolors <- c(colours, "black")
  names(mycolors) <- c("DOWN", "UP", "NO", "rbcLXS")
  p <- p + scale_colour_manual(values = mycolors)
  
  # Add vertical lines for log2FoldChange thresholds, and one horizontal line for the p-value threshold 
  if(lines_padj_FC){
    p <- p + geom_vline(xintercept=c(-foldchange, foldchange), linetype="dotted") +
      geom_hline(yintercept=-log10(padjusted), linetype="dotted") # http://www.sthda.com/english/wiki/ggplot2-line-types-how-to-change-line-types-of-a-graph-in-r-software
  }
  
  # Now write down the name of genes beside the points...
  # Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
  if(text){
    p <- p + geom_text_repel(fontface="italic")   #geom_text()
  }
  
  return(p)
}
```

```{r, warning=FALSE}
small_height <- 12
small_width <- 16
xlims <- c(-6,5.0)
ylims <- c(0,6)
text_toggle <- TRUE
p <- volcanoPlot_ggplot(highCO2_0_4, text=text_toggle, colours=c("#199e77ff", "#da5f02ff", "#d3d3d3b2"))
p <- p + xlim(xlims) + ylim(ylims)
p
ggsave("Figures/Volcano_highCO2_gen0-4_vers1.pdf", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_highCO2_gen0-4_vers2.pdf", plot=p, width=small_width, height=small_height, units="cm")
ggsave("Figures/Volcano_highCO2_gen0-4_vers1.png", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_highCO2_gen0-4_vers2.png", plot=p, width=small_width, height=small_height, units="cm")
nrow(subset(highCO2_0_4, highCO2_0_4$adj.pvalue < 0.01 & highCO2_0_4$log2FC > 1))
nrow(subset(highCO2_0_4, highCO2_0_4$adj.pvalue < 0.01 & highCO2_0_4$log2FC < -1))

p <- volcanoPlot_ggplot(highCO2_0_8, text=text_toggle, colours=c("#199e77ff", "#da5f02ff", "#d3d3d3b2"))
p <- p + xlim(xlims) + ylim(ylims)
p
ggsave("Figures/Volcano_highCO2_gen0-8_vers1.pdf", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_highCO2_gen0-8_vers2.pdf", plot=p, width=small_width, height=small_height, units="cm")
ggsave("Figures/Volcano_highCO2_gen0-8_vers1.png", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_highCO2_gen0-8_vers2.png", plot=p, width=small_width, height=small_height, units="cm")
nrow(subset(highCO2_0_8, highCO2_0_8$adj.pvalue < 0.01 & highCO2_0_4$log2FC > 1))
nrow(subset(highCO2_0_8, highCO2_0_8$adj.pvalue < 0.01 & highCO2_0_4$log2FC < -1))

p <- volcanoPlot_ggplot(lowCO2_0_4, text=text_toggle, colours=c("#199e77ff", "#da5f02ff", "#d3d3d3b2"))
p <- p + xlim(xlims) + ylim(ylims)
p
ggsave("Figures/Volcano_lowCO2_gen0-4_vers1.pdf", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_lowCO2_gen0-4_vers2.pdf", plot=p, width=small_width, height=small_height, units="cm")
ggsave("Figures/Volcano_lowCO2_gen0-4_vers1.png", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_lowCO2_gen0-4_vers2.png", plot=p, width=small_width, height=small_height, units="cm")
nrow(subset(lowCO2_0_4, lowCO2_0_4$adj.pvalue < 0.01 & highCO2_0_4$log2FC > 1))
nrow(subset(lowCO2_0_4, lowCO2_0_4$adj.pvalue < 0.01 & highCO2_0_4$log2FC < -1))

p <- volcanoPlot_ggplot(lowCO2_0_8, text=text_toggle, colours=c("#199e77ff", "#da5f02ff", "#d3d3d3b2"))
p <- p + xlim(xlims) + ylim(ylims)
p
ggsave("Figures/Volcano_lowCO2_gen0-8_vers1.pdf", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_lowCO2_gen0-8_vers2.pdf", plot=p, width=small_width, height=small_height, units="cm")
ggsave("Figures/Volcano_lowCO2_gen0-8_vers1.png", plot=p, width=24, height=18, units="cm")
ggsave("Figures/Volcano_lowCO2_gen0-8_vers2.png", plot=p, width=small_width, height=small_height, units="cm")
nrow(subset(lowCO2_0_8, lowCO2_0_8$adj.pvalue < 0.01 & highCO2_0_4$log2FC > 1))
nrow(subset(lowCO2_0_8, lowCO2_0_8$adj.pvalue < 0.01 & highCO2_0_4$log2FC < -1))
```

# Extract genes which might be affected by 2PG

```{r}
ndhR_regulon <- c("sll1594", "sll1732", "sll1733", "sll1734", "sll1735", "slr1512", "slr1513", "slr2006", "slr2007", "slr2008", "slr2009", "slr2010", "ssr3409", "ssr3410", "slr2011", "slr2013", "slr2013", "sll0834", "sll0529", "slr1592")
ndhR_regulon_proteinID <- subset(protein_slr_ID, protein_slr_ID$locus_tag %in% ndhR_regulon)$Entry

cmpr_regulon <- c("slr0040", "slr0041", "slr0042", "slr0043", "slr0044")
cmpr_regulon_proteinID <- subset(protein_slr_ID, protein_slr_ID$locus_tag %in% cmpr_regulon)$Entry

carboxysome <- c("slr1911", "sll1029", "sll1028", "slr1838", "slr1839", "sll1031", "sll1032", "sll1030", "slr0436", "slr0169")
carboxysome_proteinID <- subset(protein_slr_ID, protein_slr_ID$locus_tag %in% carboxysome)$Entry

cyabrB2_regulon <- c("sll1594", "sll0030")
cyabrB2_regulon_proteinID <- subset(protein_slr_ID, protein_slr_ID$locus_tag %in% cyabrB2_regulon)$Entry

rubisco <- c("slr0009", "slr0012")
rubisco_proteinID <- subset(protein_slr_ID, protein_slr_ID$locus_tag %in% rubisco)$Entry

ccaA <- "slr1347"
ccaA_proteinID <- subset(protein_slr_ID, protein_slr_ID$locus_tag %in% ccaA)$Entry

regulons_uniprot <- c(ndhR_regulon_proteinID, cmpr_regulon_proteinID, carboxysome_proteinID, rubisco_proteinID, cyabrB2_regulon_proteinID, ccaA_proteinID)
```

```{r}
CCM_subset_comp1_uniprot <- subset(highCO2_0_4, highCO2_0_4$Entry %in% regulons_uniprot) 
unique(CCM_subset_comp1_uniprot$Entry)
CCM_subset_comp2_uniprot <- subset(highCO2_0_8, highCO2_0_8$Entry %in% regulons_uniprot)
unique(CCM_subset_comp2_uniprot$Entry)
```

Create table etc. for that: 

```{r}
columns_of_interest <- c("Label", "log2FC", "adj.pvalue", "Entry", "gene_names_ordered_locus")
CCM_subset_comp1_uniprot <- CCM_subset_comp1_uniprot[,columns_of_interest]
CCM_subset_comp1_uniprot <- CCM_subset_comp1_uniprot %>% left_join(protein_slr_ID)
CCM_subset_comp2_uniprot <- CCM_subset_comp2_uniprot[,columns_of_interest]
CCM_subset_comp2_uniprot <- CCM_subset_comp2_uniprot %>% left_join(protein_slr_ID)

```

```{r add_annotation, echo=FALSE}
annotation <- read_tsv("annotation_locusTags_stand13012021.csv", show_col_types = FALSE)
annotation_2 <- annotation[,c(1,2,3)]
names(annotation_2) <- c("gene_names_ordered_locus", "Gene name","Product")
```

```{r}
wider_CCM_comp1 <- pivot_wider(CCM_subset_comp1_uniprot,  values_from=c(log2FC, adj.pvalue), names_from=Label)
wider_CCM_comp1 <- wider_CCM_comp1 %>% left_join(annotation_2)
write_csv(wider_CCM_comp1, "tables/CCM_subset_comparison1.csv")

#intermed <- subset(CCM_subset_comp2_uniprot, !CCM_subset_comp2_uniprot$gene_names_ordered_locus%in%c("sll1028", "sll1029", "slr0042"))
wider_CCM_comp2 <- pivot_wider(CCM_subset_comp2_uniprot, values_from=c(log2FC, adj.pvalue), names_from=Label)
wider_CCM_comp2 <- wider_CCM_comp2 %>% left_join(annotation_2)
write_csv(wider_CCM_comp2, "tables/CCM_subset_comp2.csv")
```

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
